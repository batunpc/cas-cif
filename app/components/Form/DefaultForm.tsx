interface Props {
  tableName: string;
}

const DefaultForm: React.FunctionComponent<Props> = ({ tableName }) => {
  return <div>No component was found for {tableName}.</div>;
};

export default DefaultForm;
