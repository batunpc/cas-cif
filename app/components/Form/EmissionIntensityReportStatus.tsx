import StatusBadge from "components/StatusBadge";
import {
  getDaysUntilDue,
  parseStringDate,
} from "lib/helpers/reportStatusHelpers";
import { DateTime } from "luxon";
import { useMemo } from "react";

interface Props {
  reportDueDateString?: string;
  submittedDateString?: string;
  emissionIntensityEndDateString?: string;
}

export const EmissionIntensityReportStatus: React.FC<Props> = (props) => {
  const getBadgeForEmissionIntensityReportStatus = useMemo(() => {
    const {
      reportDueDateString,
      submittedDateString,
      emissionIntensityEndDateString,
    } = props;

    const parsedDueDate = parseStringDate(reportDueDateString);
    const parsedSubmittedDate = parseStringDate(submittedDateString);
    const parsedEmissionIntensityEndDate = parseStringDate(
      emissionIntensityEndDateString
    );

    // comparing today with TEIMP end date
    const isEmissionIntensityReportEnded =
      getDaysUntilDue(parsedEmissionIntensityEndDate) < 0;

    if (parsedSubmittedDate) {
      {
        return (
          <StatusBadge
            variant="complete"
            label={`Complete (${parsedSubmittedDate.toLocaleString(
              DateTime.DATE_MED
            )})`}
          />
        );
      }
    } else if (
      parsedEmissionIntensityEndDate &&
      (!isEmissionIntensityReportEnded || !parsedDueDate)
    ) {
      return <StatusBadge variant="notDue" />;
    } else if (isEmissionIntensityReportEnded && parsedDueDate) {
      const dueIn = Math.ceil(getDaysUntilDue(parsedDueDate));
      if (dueIn < 0) return <StatusBadge variant="late" />;
      if (dueIn > 60) {
        const numberOfWeeks = Math.floor(dueIn / 7);
        return (
          <StatusBadge
            variant="dueIn"
            label={`Due in ${numberOfWeeks} ${
              numberOfWeeks > 1 ? "weeks" : "week"
            }`}
          />
        );
      }
      return (
        <StatusBadge
          variant="dueIn"
          label={`Due in ${dueIn} ${dueIn > 1 ? "days" : "day"}`}
        />
      );
    } else {
      return <StatusBadge variant="none" />;
    }
  }, [props]);
  return (
    <>
      <h3>Status</h3>
      <div className="status">
        Status of TEIMP Reporting
        {getBadgeForEmissionIntensityReportStatus}
      </div>
      <style jsx>{`
        .status {
          display: flex;
          gap: 0.5rem;
          align-items: center;
          margin-bottom: 1rem;
        }
      `}</style>
    </>
  );
};
