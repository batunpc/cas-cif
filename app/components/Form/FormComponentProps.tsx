import type { AjvError } from "@rjsf/core";
export default interface FormComponentProps {
  formData: any;
  onChange: (changeObject: object) => void;
  onFormErrors: (errorsArray: AjvError[]) => void;
}
