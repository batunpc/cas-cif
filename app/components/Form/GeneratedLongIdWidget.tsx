import {useMemo} from 'react';
import { WidgetProps } from "@rjsf/core";
import Input from "@button-inc/bcgov-theme/Input";
import Grid from '@button-inc/bcgov-theme/Grid';
import getRequiredLabel from "lib/theme/utils/getRequiredLabel";
import { graphql, useFragment } from "react-relay";

const TextWidget: React.FC<WidgetProps> = ({
  id,
  placeholder,
  onChange,
  label,
  value,
  required,
  formContext
}) => {
  const { allFundingStreamRfps } = useFragment(
    graphql`
      fragment GeneratedLongIdWidget_query on Query {
        allFundingStreamRfps {
          edges {
            node {
              rowId
              year
              fundingStreamId
            }
          }
        }
      }
    `,
    formContext.query
  );

  let fundingStreamRfp = useMemo(() => {
    return allFundingStreamRfps.edges.find(
      ({ node }) => node.rowId === formContext.form.fundingStreamRfpId
    );
  }, [allFundingStreamRfps.edges, formContext.form.fundingStreamRfpId]);

  return (
    <>
      <label htmlFor={id}>{getRequiredLabel(label, required)}</label>
      <small>&nbsp;(year - RFP - stream id - random digits - operator code)</small>
      <Grid>
        <Grid.Row style={{alignItems: 'center'}}>
            {fundingStreamRfp?.node?.year || '(YEAR)'}&nbsp;-&nbsp;RFP&nbsp;-&nbsp;{fundingStreamRfp?.node?.fundingStreamId || '(STREAM ID)'}&nbsp;-&nbsp;
          <span className='random-digits'>
            <Input
              id={id}
              onChange={(e) => onChange(e.target.value || undefined)}
              placeholder={placeholder}
              value={value || ""}
              size={"medium"}
              required={required}
            />
            </span>
            &nbsp;-&nbsp;{formContext.operatorCode || '(OPERATOR CODE)'}
        </Grid.Row>
      </Grid>
      <style jsx>
        {`
          .random-digits {
            width: 4em;
          }
        `}
      </style>
    </>
  );
};

export default TextWidget;

