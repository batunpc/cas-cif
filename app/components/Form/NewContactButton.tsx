import { Button } from "@button-inc/bcgov-theme";
import { faUserPlus } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { useCreateNewContactFormChange } from "mutations/Contact/createNewContactFormChange";
import { useAddContactToRevision } from "mutations/ProjectContact/addContactToRevision";
import { useRouter } from "next/router";
import { getContactFormPageRoute } from "pageRoutes";
import { createNewContactFormChangeMutation$data } from "__generated__/createNewContactFormChangeMutation.graphql";

interface Props {
  projectRevisionRowId: number;
  connectionString: string;
  projectContactFormId: string;
  projectId: number;
  contactIndex: number;
}

const NewContactButton: React.FC<Props> = ({
  projectRevisionRowId,
  connectionString,
  projectContactFormId,
  projectId,
  contactIndex,
}) => {
  const router = useRouter();
  const [addContact, isAddingContact] = useCreateNewContactFormChange();
  const [addContactMutation] = useAddContactToRevision();

  const clickHandler = (e: React.ChangeEvent<{}>) => {
    e.preventDefault();

    // Function to create a new contact form change and redirect to it
    const addContactFn = (formId) =>
      addContact({
        variables: {},
        onCompleted: (response: createNewContactFormChangeMutation$data) => {
          router.push(
            getContactFormPageRoute(
              response.createFormChange.formChange.id,
              formId,
              projectId,
              contactIndex
            )
          );
        },
      });

    if (projectContactFormId) {
      addContactFn(projectContactFormId);
    } else {
      // If we don't have the projectContactFormId(means we don't have the primary contact form either) we need to create one
      addContactMutation({
        variables: {
          input: {
            revisionId: projectRevisionRowId,
            contactIndex: contactIndex as number,
          },
          connections: [connectionString],
        },
        onCompleted: (res) =>
          // Creating a new contact form and redirecting to it
          addContactFn(res.addContactToRevision.formChangeEdge.node.id),
      });
    }
  };

  return (
    <Button
      onClick={clickHandler}
      style={{ margin: 0 }}
      size="medium"
      variant="primary"
      disabled={isAddingContact}
    >
      <FontAwesomeIcon icon={faUserPlus} /> Create new contact
    </Button>
  );
};

export default NewContactButton;
