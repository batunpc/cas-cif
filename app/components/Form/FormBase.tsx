import Form from "lib/theme/FormWithTheme";
import type { JSONSchema7 } from "json-schema";
import FormComponentProps from "./FormComponentProps";
interface Props extends FormComponentProps {
  schema: JSONSchema7;
  uiSchema: {};
}

const FormBase: React.FC<Props> = ({
  formData,
  onChange,
  onFormErrors,
  schema,
  uiSchema,
}) => {
  return (
    <>
      <Form
        noHtml5Validate
        schema={schema}
        uiSchema={uiSchema}
        formData={formData}
        onChange={(change) => {
          onChange(change.formData);
          onFormErrors(change.errors);
        }}
        omitExtraData
      ></Form>
      <style jsx>{`
        :global(label) {
          font-weight: bold;
        }
      `}</style>
    </>
  );
};

export default FormBase;
