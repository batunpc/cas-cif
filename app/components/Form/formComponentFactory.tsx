import DefaultForm from "./DefaultForm";
import FormComponentProps from "./FormComponentProps";
import ProjectForm from "./ProjectForm";
import ProjectManagerForm from "./ProjectManagerForm";

const staticFactory = new Map<
  string,
  React.FunctionComponent<FormComponentProps>
>([
  ["project", (props) => <ProjectForm {...props} />],
  ["project_manager", (props) => <ProjectManagerForm {...props} />],
]);

const createFormFor = (table: string) => {
  return {
    FormComponent: staticFactory.has(table)
      ? staticFactory.get(table)
      : DefaultForm,
  };
};

const formComponentFactory = { createFormFor };

export default formComponentFactory;
