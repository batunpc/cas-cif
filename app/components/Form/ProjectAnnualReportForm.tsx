import { Button } from "@button-inc/bcgov-theme";
import { faPlusCircle } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import CollapsibleReport from "components/ReportingRequirement/CollapsibleReport";
import ReportDueIndicator from "components/ReportingRequirement/ReportDueIndicator";
import Status from "components/ReportingRequirement/Status";
import projectReportingRequirementSchema from "data/jsonSchemaForm/projectReportingRequirementSchema";
import { JSONSchema7 } from "json-schema";
import { getReportingStatus, isOverdue } from "lib/helpers/reportStatusHelpers";
import FormBorder from "lib/theme/components/FormBorder";
import EmptyObjectFieldTemplate from "lib/theme/EmptyObjectFieldTemplate";
import { useAddReportingRequirementToRevision } from "mutations/ProjectReportingRequirement/addReportingRequirementToRevision";
import useDiscardReportingRequirementFormChange from "mutations/ProjectReportingRequirement/discardReportingRequirementFormChange";
import { useUpdateReportingRequirementFormChange } from "mutations/ProjectReportingRequirement/updateReportingRequirementFormChange";
import { MutableRefObject, useMemo, useRef } from "react";
import { graphql, useFragment } from "react-relay";
import { ProjectAnnualReportForm_projectRevision$key } from "__generated__/ProjectAnnualReportForm_projectRevision.graphql";
import FormBase from "./FormBase";
import {
  addReportFormChange,
  deleteReportFormChange,
  getSortedReports,
  stageReportFormChanges,
  updateReportFormChange,
} from "./reportingRequirementFormChangeFunctions";
import SavingIndicator from "./SavingIndicator";
import UndoChangesButton from "./UndoChangesButton";

interface Props {
  onSubmit: () => void;
  projectRevision: ProjectAnnualReportForm_projectRevision$key;
}

export const annualReportUiSchema = {
  reportDueDate: {
    "ui:widget": "DueDateWidget",
  },
  submittedDate: {
    "ui:widget": "ReceivedDateWidget",
  },
  comments: {
    "ui:widget": "TextAreaWidget",
  },
};

const ProjectAnnualReportForm: React.FC<Props> = (props) => {
  const formRefs: MutableRefObject<{}> = useRef({});

  const projectRevision = useFragment(
    graphql`
      fragment ProjectAnnualReportForm_projectRevision on ProjectRevision {
        id
        # eslint-disable-next-line relay/unused-fields
        rowId
        projectAnnualReportFormChanges: formChangesFor(
          formDataTableName: "reporting_requirement"
          reportType: "Annual"
          first: 502
        ) @connection(key: "connection_projectAnnualReportFormChanges") {
          __id
          edges {
            # eslint-disable-next-line relay/unused-fields
            node {
              rowId
              id
              newFormData
              operation
              changeStatus
              formChangeByPreviousFormChangeId {
                changeStatus
                newFormData
              }
              asReportingRequirement {
                ...CollapsibleReport_reportingRequirement
              }
            }
          }
        }

        upcomingAnnualReportFormChange: upcomingReportingRequirementFormChange(
          reportType: "Annual"
        ) {
          # eslint-disable-next-line relay/must-colocate-fragment-spreads
          ...ReportDueIndicator_formChange
          asReportingRequirement {
            reportDueDate
          }
        }
        # eslint-disable-next-line relay/unused-fields
        projectFormChange {
          formDataRecordId
        }
      }
    `,
    props.projectRevision
  );

  const [addAnnualReportMutation, isAdding] =
    useAddReportingRequirementToRevision();

  const [applyUpdateFormChangeMutation, isUpdating] =
    useUpdateReportingRequirementFormChange();

  const [discardFormChange] = useDiscardReportingRequirementFormChange(
    "Annual",
    projectRevision.projectAnnualReportFormChanges.__id
  );

  const [sortedAnnualReports, nextAnnualReportIndex] = useMemo(() => {
    return getSortedReports(
      projectRevision.projectAnnualReportFormChanges.edges
    );
  }, [projectRevision.projectAnnualReportFormChanges]);

  // Get all form changes ids to get used in the undo changes button
  const formChangeIds = useMemo(() => {
    return projectRevision.projectAnnualReportFormChanges.edges.map(
      ({ node }) => node?.rowId
    );
  }, [projectRevision.projectAnnualReportFormChanges]);

  const reportDueDate =
    projectRevision.upcomingAnnualReportFormChange?.asReportingRequirement
      .reportDueDate;
  const overdue = useMemo(() => {
    return isOverdue(reportDueDate);
  }, [reportDueDate]);

  const reportSubmittedDates = useMemo(() => {
    return projectRevision.projectAnnualReportFormChanges.edges.map(
      ({ node }) => node.newFormData.submittedDate
    );
  }, [projectRevision.projectAnnualReportFormChanges.edges]);

  const variant = getReportingStatus(reportSubmittedDates, overdue);

  return (
    <div>
      <header>
        <h2>Annual Reports</h2>
        <UndoChangesButton formChangeIds={formChangeIds} formRefs={formRefs} />
        <SavingIndicator isSaved={!isUpdating && !isAdding} />
      </header>
      <h3>Status</h3>
      <Status variant={variant} reportType={"Annual"} />

      <ReportDueIndicator
        reportTitle="Annual Report"
        reportDueFormChange={projectRevision.upcomingAnnualReportFormChange}
      />
      <FormBorder>
        <Button
          variant="secondary"
          onClick={() =>
            addReportFormChange(
              addAnnualReportMutation,
              projectRevision,
              nextAnnualReportIndex,
              "Annual",
              projectRevision.projectAnnualReportFormChanges.__id
            )
          }
          className="addButton"
        >
          <FontAwesomeIcon icon={faPlusCircle} /> Add another annual report
        </Button>

        {sortedAnnualReports.map((report, index) => {
          return (
            <div key={report.id}>
              <CollapsibleReport
                title={`Annual Report ${index + 1}`}
                reportingRequirement={report.asReportingRequirement}
              >
                <Button
                  variant="secondary"
                  size="small"
                  onClick={() =>
                    deleteReportFormChange(
                      discardFormChange,
                      report.id,
                      report.operation,
                      formRefs
                    )
                  }
                  className="removeButton"
                >
                  Remove
                </Button>
                <FormBase
                  id={`form-${report.id}`}
                  validateOnMount={report.changeStatus === "staged"}
                  idPrefix={`form-${report.id}`}
                  ref={(el) => (formRefs.current[report.id] = el)}
                  formData={report.newFormData}
                  onChange={(change) => {
                    updateReportFormChange(
                      applyUpdateFormChangeMutation,
                      "Annual",
                      { ...report, changeStatus: "pending" },
                      change.formData
                    );
                  }}
                  schema={projectReportingRequirementSchema as JSONSchema7}
                  uiSchema={annualReportUiSchema}
                  ObjectFieldTemplate={EmptyObjectFieldTemplate}
                  formContext={{
                    dueDate: report.newFormData?.reportDueDate,
                  }}
                />
              </CollapsibleReport>
            </div>
          );
        })}
      </FormBorder>
      <Button
        size="medium"
        variant="primary"
        onClick={() =>
          stageReportFormChanges(
            applyUpdateFormChangeMutation,
            "Annual",
            props.onSubmit,
            formRefs,
            projectRevision.projectAnnualReportFormChanges.edges
          )
        }
        disabled={isUpdating}
      >
        Submit Annual Reports
      </Button>
      <style jsx>{`
        div :global(button.pg-button) {
          margin-left: 0.4em;
          margin-right: 0em;
        }
        div :global(button.addButton) {
          margin-bottom: 1em;
        }
        div :global(button.removeButton) {
          float: right;
        }
      `}</style>
    </div>
  );
};

export default ProjectAnnualReportForm;
