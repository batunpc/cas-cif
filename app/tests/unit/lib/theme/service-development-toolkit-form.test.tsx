import React from "react";

import { render } from "@testing-library/react";
import Form from "lib/theme/service-development-toolkit-form";
import { JSONSchema7 } from "json-schema";

describe("The Form", () => {
  it("Renders a basic test schema with a submit button", () => {
    const testSchema: JSONSchema7 = {
      type: "object",
      properties: {
        test_number: { type: "number", title: "Test Number" },
        test_string: { type: "string", title: "Test Script" },
      },
    };

    const componentUnderTest = render(<Form schema={testSchema} />);

    expect(componentUnderTest.container).toMatchSnapshot();
  });
});
