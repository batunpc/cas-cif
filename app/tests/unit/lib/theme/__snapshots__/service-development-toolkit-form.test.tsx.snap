// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Form Renders a basic test schema with a submit button 1`] = `
<div>
  <form
    class="rjsf"
  >
    <fieldset
      class="sc-dkPtRN bfJYwN pg-fieldset"
      id="root"
    >
      <div
        class="sc-hBUSln jEYZxL pg-input"
      >
        <label
          class="sc-fotOHu bhxnix pg-input-label"
          for="root_test_number"
        >
          Test Number
        </label>
        <input
          aria-label="Test Number"
          class="sc-fFeiMQ jSxRsG pg-input-input"
          id="root_test_number"
          name="root_test_number-input"
          placeholder=""
          value=""
        />
      </div>
      <div
        class="sc-hBUSln jEYZxL pg-input"
      >
        <label
          class="sc-fotOHu bhxnix pg-input-label"
          for="root_test_string"
        >
          Test Script
        </label>
        <input
          aria-label="Test Script"
          class="sc-fFeiMQ jSxRsG pg-input-input"
          id="root_test_string"
          name="root_test_string-input"
          placeholder=""
          value=""
        />
      </div>
    </fieldset>
  </form>
</div>
`;
